src/api/app/models/event_subscription/generate_hash_for_subscriber.rb

```diff
@@ -0,0 +1,72 @@
+class EventSubscription
+  class GenerateHashForSubscriber
+    attr_reader :subscriber
+
+    def initialize(subscriber)
+      @subscriber = subscriber
+    end
+
+    # Generate a hash of subscriptions for the user grouped by the subscription's event class
+    def query
+      event_classes = Event::Base.notification_events
+      subscriptions_by_event_class = {}
+
+      event_classes.each do |event_class|
+        event_class.receiver_roles.each do |role|
+          subscription = find_subscription_for_event_class_and_role(event_class, role)
+
+          subscriptions_by_event_class[event_class] ||= []
+          subscriptions_by_event_class[event_class] << subscription
+        end
+      end
+
+      subscriptions_by_event_class
+    end
+
+    private
+
+    def find_subscription_for_event_class_and_role(event_class, role)
+      subscriber_subscription = find_subscriber_subscription(event_class, role)
+      default_subscription = find_default_subscription(event_class, role)
+
+      # 1. Pick the subscriber's subscription if it exists
+      if subscriber.present? && subscriber_subscription.present?
+        subscriber_subscription
+
+      # 2. Pick the subscriber's subscription if it exists
+      elsif default_subscription.present?
+        default_subscription
+
+      # 3. Otherwise instantiate a new subscription
+      else
+        EventSubscription.new(
+          subscriber: subscriber,
+          eventtype: event_class.to_s,
+          receiver_role: role,
+          receive: false
+        )
+      end
+    end
+
+    def find_subscriber_subscription(event_class, role)
+      subscription =
+        subscriber_subscriptions.find { |s| s.event_class == event_class && s.receiver_role == role }
+
+      return subscription if subscription.present?
+
+      subscriber_subscriptions.find { |s| s.event_class == event_class && s.receiver_role == :all }
+    end
```

Comment:
Untested, but this might work as well. And it seems easier to read / understand to me.

Might be even possible to replace the private method by a scope.

```ruby
def find_subscriber_subscription(event_class, role)
  find_subscriber_by_role(role) || find_subscriber_by_role(:all)
end

private

def find_subscriber_by_role(role)
  subscriber_subscriptions.find { |s| s.event_class == event_class && s.receiver_role == role }
end
```